fix(materialized columns): Create materialized columns without dollar sign in column name

macobo opened the pull request:
## Changes

Follow-up to https://github.com/PostHog/posthog/pull/8846

Our materialization logic currently creates columns with a $ sign. According to the linked PR, clickhouse added heredoc support, breaking this feature.

This PR introduces an alternative solution by removing $ from the column names. I'd love to roll back the linked PR and use this instead as I think this has less of a chance of becoming a frequent footgun during development.

## How did you test this code?

See tests. Also ran the migration on an existing install as well as a fresh one.

Planning on manually testing the migration against prod before merging.

macobo committed:
Get rid of $ signs in column names

Clickhouse introduced heredocs which made $ signs backwards incompatible

macobo committed:
Migration to rename existing materialized columns

macobo commented on ee/clickhouse/models/event.py:
```diff
@@ -228,7 +228,7 @@ def get_agg_events_with_groups_count_for_teams_and_period(
         FROM events
         WHERE team_id IN (%(team_id_clause)s)
         AND timestamp between %(begin)s AND %(end)s
-        AND ($group_0 != '' OR $group_1 != '' OR $group_2 != '' OR $group_3 != '' OR $group_4 != '')
```
Note: This was a spot missed in the previous PR

hazzadous replied:
Yes I did just enough to fix the tests an no more, leaving that to a separate followup.

hazzadous commented:
> I'd love to roll back the linked PR and use this instead as I think this has less of a chance of becoming a frequent footgun during development

I'd suggest we do both, if we can ensure that all identifiers are quoted then that should remove some footguns, coupled with only using boring characters in identifier names.

macobo committed:
Update some snapshots

macobo committed:
Remove extra_event_properties - SQL injection risk

macobo committed:
Update snapshots

macobo commented on ee/clickhouse/queries/paths/path_event_query.py:
```diff
@@ -45,11 +45,6 @@ def get_query(self) -> Tuple[str, Dict[str, Any]]:
-        _fields += [
-            get_property_string_expr("events", field, f"'{field}'", "properties", table_alias=self.EVENT_TABLE_ALIAS)[0]
-            + f" as {field}"
-            for field in self._extra_event_properties
```
Also needed to clean this up due to a potential SQL injection risk/usage of $ in column names.

macobo committed:
Update DISTINCT group count

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,48 @@
+def rename_column_if_exists(table, current_name, new_name):
+    try:
+        rename_column(table, current_name, new_name)
+    except ServerException as err:
```
Is there any follow up needed in this case, for the operator?

macobo replied:
Nope - it's already renamed, can ignore.

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,48 @@
+    try:
+        rename_column(table, current_name, new_name)
+    except ServerException as err:
+        logger.info("Failed to rename column, ignoring.", table=table, current_name=current_name, new_name=new_name)
```
This seems more like a warning rather than info. Also, can we capture the exception with `exc_info=True`.

macobo replied:
Will clear up the wording. This is safe to ignore as the column would already be renamed on a fresh install.

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,48 @@
+        logger.info("Failed to rename column, ignoring.", table=table, current_name=current_name, new_name=new_name)
+
+
+def rename_materialized_columns_with_dollars(database):
```
What happens while this is running, or if we fail part way through for some reason? If old code is still trying to access $group will that be an issue?

macobo replied:
Good question.

We won't fail part way through, but old code will be running even after this is complete. Due to caching a few queries will fail as this is deployed as a result until the new pods are live. As such, I'd deploy this during a quieter timeframe.

hazzadous reviewed (approved):
Looks good, a bit hazy on the exact details.

Looks like there was still a reference to $ in `ee/clickhouse/queries/test/__snapshots__/test_stickiness.ambr` when I checked, maybe this is a stale snapshot?

macobo committed:
Remove old snapshot file

macobo committed:
Update log message

rcmarron commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,48 @@
+                continue
+
+            new_column_name = materialized_column_name(table, property)  # type: ignore
+            rename_column(table, column_name, new_column_name)
```
Flyby comment: if a self-hosted customer had `$mat_session_id` materialized on their instance, could this fail bc we're trying to rename `$mat_session_id` to `mat_session_id` which just got created above?

macobo replied:
Might be because I guess the migration creating the column `mat_session_id` broke the invariant that one property is only materialized once :(

macobo replied:
Nevermind, we validate that invariant in code.

macobo committed:
Update ON CLUSTER stmts

guidoiaquinti reviewed (approved):
Merge conflict but overall LGTM

hazzadous commented:
@macobo to let you get on with other tasks did you want to pass this resolution along to someone else? Is there anything else to do here other than resolve conflicts?

macobo commented:
> @macobo to let you get on with other tasks did you want to pass this resolution along to someone else? Is there anything else to do here other than resolve conflicts?

Nope, I will get this merged. This was originally blocked by the trouble we were having on ch8, now want to make sure it works well with sharded setups and depends on https://github.com/PostHog/posthog/pull/9013.

macobo committed:
Update snapshots

macobo committed:
Minor fixes

macobo committed:
Fix column names

macobo committed:
Update some snapshots

macobo committed:
Update imports

macobo commented:
Deploying first thing tomorrow...

macobo committed:
ALIAS not rename in dollar sign migration

On production we ran into issues renaming a bunch of columns, resulting
in a deadlocked cluster. Creating alias columns on distributed tables
fixes this issue and seems to work well

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,74 @@
+    )
+
+
+def alias_column_if_exists(table, current_name, new_name):
```
We do this if and the IF NOT EXIST SQL to avoid hitting the db all the time, is that right?

macobo replied:
IF NOT EXISTS is for checking if the _new_ column exists, this method is for checking whether the _old_ column also exists (which it won't on new setups)

hazzadous replied:
Ah I see it's for the current name, ignore

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,74 @@
+        f"""
+        ALTER TABLE {table} ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+        ADD COLUMN IF NOT EXISTS \"{new_name}\"
+        VARCHAR ALIAS \"{current_name}\"
```
Is it always varchar? I guess we treat everything as a string then interpret on read if we need to?

macobo replied:
Materialized columns are all varchars currently yeah. :)

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,74 @@
+
+    comment = get_column_names(table)[current_name]
+    if len(comment) > 0:
+        comment_column(table, current_name, "")
```
If we fail for any reason on the second comment, I guess we would not used a materialized column for some queries. Is it worth switching the order here such that we always maintain at least one materialized column comment (I have no idea what the behaviour here is however, but maybe it just uses the first one it comes across?)

I assume also that it will use the cached materialized column names also, but that's fine because the old name still exists. I like this solution

macobo replied:
> Is it worth switching the order here such that we always maintain at least one materialized column comment 

I am not sure how code would behave if some comments are duplicated - I think it'll be undeterministic. Losing columns sounds better this way.

You're right in that this migration isn't really rollbackable, but it should be fine as prod is currently the only sharded cluster if we take care there.

hazzadous commented on ee/clickhouse/migrations/0025_remove_dollar_from_materialized_columns.py:
```diff
@@ -0,0 +1,74 @@
+            if "$" not in column_name:
+                continue
+
+            new_column_name = "alias_" + materialized_column_name(table, property)  # type: ignore
```
I assume `materialized_column_name` never returns something with a `$` because of https://github.com/PostHog/posthog/pull/8855/files#diff-559ae24ad49a1674573c0c89db08c7b6d96c0920e2625e0877b2155f0f453aa1L141

And that is tested by https://github.com/PostHog/posthog/pull/8855/files#diff-6653caa40d66302b910d83e95b18f8176e45400c230eed900e5dc7ef222577ceL91 ?

hazzadous reviewed (approved):
Looks good 🚀 

I didn't see a test for this but I may have missed. I would love to know that the code has at least run before trying out live, but I leave that to your discretion.

macobo commented:
> I would love to know that the code has at least run before trying out live, but I leave that to your discretion.

It has! Locally ran migrations on master, created a materialized column, switched here, ran 0025 and smoke-tested querying the old and new columns + what get_materialized_columns returns.

macobo committed:
Fix on cluster

fuziontech commented:
Revived!

fuziontech committed:
Merge branch 'master' into materialized-columns-without-dollar

* master: (302 commits)
  chore(styling): remove mixin-elevated (#9644)
  Fix cors sentry (#9642)
  fix(capture): CORS error with sentry browser tracing (#9639)
  feat(plugin-server): allow historical exports to be enabled and disabled via env (#9606)
  refactor(scenes): automatically bind props to logic in scenes (#9451)
  fix(cohort): params not stringified (#9632)
  fix(onboarding-1.0): blue buttons and small tweaks (#9615)
  ci(zookeeper version/image): use zookeeper:3.7.0 (#9638)
  feat(funnels): Add tooltip to new funnel viz (#9583)
  chore(ci): migrate husky from v4->v7 (#9633)
  fix: update latest migrations file (#9622)
  fix(dashboards): Allow sorting only a specific table on a dashboard (#9625)
  fix: recalculating cohort on update (#9630)
  fix(cohort): remove api check temp (#9628)
  perf(ingestion): stop consuming from buffer for good (for now) (#9626)
  feat: move insights between dashboards when many-to-many (#9598)
  fix: no need to raise an error if cache key has no results (#9619)
  chore(cohorts): Add acceptance criteria for new cohorts and feature flag functionality (#9624)
  feat(person): Added setting for customising the displayed property for the Person identifier (#9607)
  chore(cohorts): Revert Add acceptance criteria for new cohorts and feature flag functionality (#9623)
  ...

fuziontech committed:
cleaning up merge (was very sloppy)

fuziontech committed:
more sanity checks on the merge

fuziontech committed:
black format

fuziontech committed:
flake8

fuziontech committed:
clean up amber files

fuziontech committed:
more amber spot checks

fuziontech committed:
more fixes for tests

fuziontech committed:
more cleanup

fuziontech committed:
a bit more fix'n

fuziontech commented:
Closing in favor of https://github.com/PostHog/posthog/pull/9721
Thanks @rcmarron !
